### PR TITLE
Re-enable the 'public code' box in the Manage Repositories page, & hook up to graphql APIs

### PR DESCRIPTION
**Pull request by arussellsaw**
This PR hooks up the SetUserPublicRepos graphql API to the manage repositories page, allowing users to list public repos they're interested in searching, and have those repos appear in their searches.

We probably want to make a few further changes:
* enable search contexts if the user doesn't have any code hosts or directly added repos
* show public repos from this box on the previous page

https://user-images.githubusercontent.com/5236823/110816841-9bbb4480-8283-11eb-9242-d878f9dc1abd.mov

### CODE REVIEW

**Comment by artemruts** [comment]
Hey @arussellsaw a few questions:

1. what will happen if user adds a string that doesn't represent a valid public repo URL (`https://example.com/bla`, `https://github.com/private/veryprivate`)? Do we expect some error or a banner shown?
2. when the used adds a valid public repo URL, should it be listed among all of the synced repos `(settings/repositories` view)?

I noticed that repos added with the `textarea` are not a part of the `graphql?UserRepositories` responses and don't count (as in `repoCount`  of repos for the given code host) in the `graphql?ExternalServices` responses, is this by design?

File: internal/database/user_public_repos.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ func (s *UserPublicRepoStore) SetUserRepos(ctx context.Context, userID int32, re
 			userID, repo.RepoURI, repo.RepoID,
 		))
 	}
+	if len(values) == 0 {
```
Comment (ryanslade) [inline internal/database/user_public_repos.go:54]: I think you can move this check to the top and instead return early if `len(repos) == 0`

File: client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx [hunk 2]
```diff
@@ -638,29 +688,37 @@ export const UserSettingsManageRepositoriesPage: React.FunctionComponent<Props>
...
-                                <textarea className="form-control" rows={5} />
-                                <p className="text-muted mt-2">Specify with complete URLs. One repository per line.</p>
-                            </div>
-                        )}
-                    </div>
-                </li>
+                {window.context.sourcegraphDotComMode && (
+                    <li className="list-group-item p-0 user-settings-repos__container" key="add-textarea">
+                        <div className="p-4 text-muted">
+                            <h3 className="text-muted">Other public repositories</h3>
```
Comment (quinnkeast) [inline client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx:694]:
```suggestion
                            <h3>Other public repositories</h3>
```

File: client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx [hunk 3]
```diff
@@ -638,29 +688,37 @@ export const UserSettingsManageRepositoriesPage: React.FunctionComponent<Props>
...
-                                <p className="mb-2">Repositories to sync</p>
-                                <textarea className="form-control" rows={5} />
-                                <p className="text-muted mt-2">Specify with complete URLs. One repository per line.</p>
-                            </div>
-                        )}
-                    </div>
-                </li>
+                {window.context.sourcegraphDotComMode && (
+                    <li className="list-group-item p-0 user-settings-repos__container" key="add-textarea">
+                        <div className="p-4 text-muted">
```
Comment (quinnkeast) [inline client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx:693]:
```suggestion
                        <div className="p-4">
```

**Review by quinnkeast — COMMENTED** [review]
Added a couple of suggestions to remove the muted style from a few text elements, so that they're fully "available" now.

>enable search contexts if the user doesn't have any code hosts or directly added repos

Yes, we should show search contexts as soon as they've "added repos," which could be either through connected code hosts or this URL option.

>show public repos from this box on the previous page

100%, any repos added via this "Manage repos" view are "repositories you've added to Sourcegraph" and should appear in the list of repositories.

I imagine we should do some validation on the provided URLs before leaving this view on save, both for format and for being a valid public repository. Would it be feasible to do this and potentially insert an error indicator for specific lines in the text input, kind of like a merge conflict?

**Comment by arussellsaw** [comment]
@artemruts @quinnkeast addessed feedback, small example of better errors along with new error placement, and seeing repos in the repo list 

https://user-images.githubusercontent.com/5236823/110932831-e7252f80-8323-11eb-8a18-d92b7fc75f7a.mov

**Review by unknwon — APPROVED** [review]
Backend change LGTM

File: client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx [hunk 4]
```diff
@@ -294,6 +325,26 @@ export const UserSettingsManageRepositoriesPage: React.FunctionComponent<Props>
...
+            if (!publicRepoState.enabled) {
+                publicRepos = []
+            }
+            let didError = false
+            await setUserPublicRepositories(userID, publicRepos)
+                .toPromise()
+                .catch(error => {
+                    setRepoState({ ...repoState, error: String(error) })
+                    didError = true
+                })
```
Comment (artemruts) [inline client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx:339]: Maybe we don't need to make a `setUserPublicRepositories` call  if the `publicRepoState.enabled` is not `true`? We could also skip splitting lines then.

**Review by artemruts — APPROVED** [review]

**Review by asdine — APPROVED** [review]

File: client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx (hunk 4 above)
Comment (arussellsaw) [reply client/web/src/user/settings/repositories/UserSettingsManageRepositoriesPage.tsx:339]: we still want to make that request, as if it's not true we need to update the list of repos to be empty

**Pull request merged**